### PR TITLE
[RANS] RANSApplication Cleanup

### PR DESCRIPTION
**Pull request by sunethwarna**
Hi,

This adds a time averaging process to RANSApplicaton, the time averaged values of historical variables are stored in their non-historical data value container. A test is also added for the process.

Some cleanup is also done in this PR.

### CODE REVIEW

File: applications/RANSApplication/custom_processes/auxiliary_processes/rans_check_vector_bounds_process.cpp [hunk 1]
```diff
@@ -113,36 +115,49 @@ void RansCheckVectorBoundsProcess::Execute()
...
+    OpenMPUtils::PartitionVector node_partition;
+    OpenMPUtils::DivideInPartitions(number_of_nodes, number_of_threads, node_partition);
+
+    Vector max_values(number_of_threads, max_value);
+    Vector min_values(number_of_threads, min_value);
 
-        double current_value = 0.0;
-        for (int dim = 0; dim < 3; ++dim)
-            current_value += std::pow(vector_value[dim] * vector_weights[dim], 2);
+#pragma omp parallel
```
Comment (philbucher) [inline applications/RANSApplication/custom_processes/auxiliary_processes/rans_check_vector_bounds_process.cpp:129]: probably better to use a [reduction](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/solving_strategies/convergencecriterias/displacement_criteria.h#L354) here ;)

Comment (loumalouomega) [reply applications/RANSApplication/custom_processes/auxiliary_processes/rans_check_vector_bounds_process.cpp:129]: Reduction is quite troublesome with VS

File: applications/RANSApplication/custom_processes/auxiliary_processes/rans_check_vector_bounds_process.cpp [hunk 2]
```diff
@@ -113,36 +115,49 @@ void RansCheckVectorBoundsProcess::Execute()
         break;
     }
 
-    double min_value = 0.0;
-    double max_value = 0.0;
-    bool is_initialized = false;
+    double min_value = std::numeric_limits<double>::max();
+    double max_value = std::numeric_limits<double>::min();
```
Comment (philbucher) [inline applications/RANSApplication/custom_processes/auxiliary_processes/rans_check_vector_bounds_process.cpp]: is it `min` or `lowest`? I am not sure, but I remember that I had an unplesant surprise some time ago

File: applications/RANSApplication/custom_processes/auxiliary_processes/rans_check_vector_bounds_process.cpp (hunk 1 above)
Comment (philbucher) [reply applications/RANSApplication/custom_processes/auxiliary_processes/rans_check_vector_bounds_process.cpp:129]: thats why I was suggesting sth that is in the core and hence works in VS
check the link ;)

Comment (sunethwarna) [reply applications/RANSApplication/custom_processes/auxiliary_processes/rans_check_vector_bounds_process.cpp:129]: Is there a way to use min, max reductions in windows with VS? I had a compilation error before, so I had to change it to this. :/

File: applications/RANSApplication/custom_processes/auxiliary_processes/rans_check_vector_bounds_process.cpp (hunk 2 above)
Comment (sunethwarna) [reply applications/RANSApplication/custom_processes/auxiliary_processes/rans_check_vector_bounds_process.cpp]: you are correct, it should be lowest. thanks :)

**Comment by riccardotosi** [comment]
Hi @sunethwarna , I am writing because inside ExaquteSandboxApplication we are also computing time averaged quantities. Currently, only the one for the elemental divergence is merged in the master, other nodal time averages are still in a branch. Don't you/we think that "rans_time_averaging_process" should be in the core?

**Comment by sunethwarna** [comment]
@riccardotosi I can move it to the core if everyone agrees. @philbucher ?

**Comment by philbucher** [comment]
is it only for rans (=> Fluid) or in general for transient problems (=> core)?

**Comment by sunethwarna** [comment]
It is for general time averaging. nothing to do with rans nor fluid. So can I move the test, and the process to the core in the same PR? or should I create a new PR after this?

**Comment by rubenzorrilla** [comment]
And why not a `TimeAveragingApplication` or `StatisticsApplication`? I wouldn't like to start putting more and more processes to the kratos core, which is quite plenty of them right now. 

However I'm not an expert on this topic so I can't foresee if you might eventually need more processes or this one can do the job for all the possible use cases. If it is just one, it would make sense to store it in the core.

BTW, is it possible that both @sunethwarna and @riccardotosi use the same implementation (or base class) to do the time averaging?

**Comment by sunethwarna** [comment]
@riccardotosi could you point me to the process you are using?, then we can compare and come-up with a common solution?

@rubenzorrilla StatisticsApplication makes sense, but for the time being I only need the time averaged mean as a statistical quantity. So if @riccardotosi or someone else have more requirements we can go for a new application ryt?

**Comment by riccardotosi** [comment]
Hi @sunethwarna , [here](https://github.com/KratosMultiphysics/Kratos/blob/exaqute/UQ_demonstrator/applications/ExaquteSandboxApplication/custom_processes/weighted_pressure_calculation_process.cpp) you can find an example of what we are using. It is not general as yours for now, reason why it is not merged. It seems to me that we are using the same implementation, meaning we calculate nodal time averages. I understand from your code that you also considered "wash-out times" to account for deleting any bias coming from initial condition.

From our side, we need more statistical processes, and they are (or will be) stored in MultilevelMonteCarloApplication, which is the application we use to do Uncertainty Quantification.

**Comment by sunethwarna** [comment]
@riccardotosi is there a possibility to store statistical process in StatisticsApplication without integrating them into specific applications or core?

**Comment by sunethwarna** [comment]
I can put this time_averaging process in the StatisticsApplication as a start.

**Comment by riccardotosi** [comment]
MultilevelMonteCarloApplication is a "non-physical" application in which utilities or python scripts required to run Uncertainty Quantification, such as Monte Carlo or power sums (which are statistical tools) for example, are stored.

I wouldn't mind creating a statistical application, but I see it as a duplicate. I do not know about other opinions.

**Comment by riccardotosi** [comment]
Hi @sunethwarna , apart from nodal weighted time averaging, which are the statistical tools you are interested in?

**Comment by sunethwarna** [comment]
At the moment, I only need nodal time weighted averaging. May be we can disscuss about a common solution after the meeting you are having with Anoop?

**Comment by riccardotosi** [comment]
Yes sure!

**Comment by sunethwarna** [comment]
I removed the time averaging process from this PR, and it is added to the core from #6078. Now this PR is merely a cleanup of the RANSApplication (@riccardotosi).

@philbucher, @rubenzorrilla could you have a look?

**Comment by philbucher** [comment]
this will be ready for review after #6078 right?

**Comment by philbucher** [comment]
ah nevermind, just red your comment

File: applications/RANSApplication/custom_processes/solving_strategies/scalar_co_solving_process.h [hunk 3]
```diff
@@ -89,6 +93,17 @@ class ScalarCoSolvingProcess : public Process
...
         mSkipIterations = rParameters["number_of_parent_solve_iterations"].GetInt();
+        mVtkOutputFrequency = rParameters["vtk_output_frequency"].GetInt();
+        mVtkOutputPrefix = rParameters["vtk_output_prefix"].GetString();
+
+        Parameters empty_parameters(R"({})");
+        if (!rParameters["vtk_output_settings"].IsEquivalentTo(empty_parameters))
+        {
+            KRATOS_INFO_IF(this->Info(), mEchoLevel > 0)
+                << "Adding VtkOutput.\n";
+            mpVtkOutput = unique_ptr<VtkOutput>(
```
Comment (philbucher) [inline applications/RANSApplication/custom_processes/solving_strategies/scalar_co_solving_process.h]:
```suggestion
            mpVtkOutput = Kratos::make_unique<VtkOutput>(
```

**Review by philbucher — DISMISSED** [review]
one suggestion

File: applications/RANSApplication/custom_processes/solving_strategies/scalar_co_solving_process.h (hunk 3 above)
Comment (philbucher) [reply applications/RANSApplication/custom_processes/solving_strategies/scalar_co_solving_process.h]: https://stackoverflow.com/questions/22571202/differences-between-stdmake-unique-and-stdunique-ptr-with-new

**Comment by sunethwarna** [comment]
@philbucher could you please re-approve?

**Review by philbucher — APPROVED** [review]
now should be fine

**Comment by sunethwarna** [comment]
Thanks :)

**Pull request merged**